Add filesystem based LLEXT loader

awojasinski opened the pull request:
This PR introduces new LLEXT loader that can load extension files from filesystem.

Correct use of file descriptors required two additional methods in `struct llext_loader`.
The `llext_prepare()` that's invoked at the beginning of llext loading procedure,
before anything is read from the extension, and the `llext_finalize()` function that is called
after the llext has been loaded and the extension content is no longer needed.

This functions can be undefined for all other loaders that don't require such functionality.
The buffer loader structure definition has been aligned to changes in `struct llext_loader`.

The patch doesn't include any test as it requires manual steps in order to place extension
in a file system. @kartben I could add description similar to
https://docs.zephyrproject.org/latest/samples/subsys/fs/littlefs/README.html describing
how to modify `samples/subsys/llext/modules` sample to load extension from file.
Unfortunately loading filesystem image is board specific for my testing on `nrf52840dk`
it required use of `bin2hex.py` and `nrfjprog` to load it into external flash.

awojasinski committed:
llext: Add new methods to loader API

Introducing `llext_prepare()` and `llext_finalize()` APIs
that are invoked and the beginning and the end of the `llext_load()`
function.

The purpose of these functions is to bring possibility of initializing
loader before it is used and uninitialize or clean up when
it is no longer needed. Both functions are optional.

The buffer loader has been aligned to methods introduced in the patch.

Signed-off-by: Adam Wojasinski <awojasinski@baylibre.com>

awojasinski committed:
llext: Add filesystem based extension loader

Added loader allowing to load extensions stored in a filesystem.

Signed-off-by: Adam Wojasinski <awojasinski@baylibre.com>

awojasinski committed:
llext: Add shell command for loading LLEXT from a filesystem

This patch extends LLEXT shell commands with support for loading
LLEXT from a filesystem. Use of the command requires absolute path
to the llext file.

Example use:
`llext load_llext hello_world /lfs/hello_world.llext`

Signed-off-by: Adam Wojasinski <awojasinski@baylibre.com>

awojasinski commented:
@teburd actually sounds like pretty good idea, I'll include that in the PR as new commit

npitre commented:
small comments:

```
+ * @brief Initializer for an llext_buf_loader structure
+ *
+ * @param _filename Buffer containing the ELF binary
```

You probably meant "llext_fs_loader structure", and `_filename` also appears
mis-defined.

pillo79 commented on subsys/llext/Kconfig:
```diff
@@ -47,6 +47,12 @@ config LLEXT_HEAP_SIZE
 	help
 	  Heap size in kilobytes available to llext for dynamic allocation
 
+config LLEXT_FILE_SYSTEM
```
Personally I don't see the need to disable LLEXT file system support if we have support for it in the rest of Zephyr. Do you have an use case in mind? Otherwise you can directly use `FILE_SYSTEM` everywhere.

awojasinski replied:
Don't have any use case in mind. I can change the feature to be based on `FILE_SYSTEM`. If anyone will need to have extra Kconfig symbol then it will be up to her/him to submit patch

pillo79 reviewed (commented):
LGTM, thanks for this! One topic I would like feedback on.

lyakh commented on subsys/llext/fs_loader.c:
```diff
@@ -0,0 +1,65 @@
+		ret = -EINVAL;
+	}
+
+	return ret == len ? 0 : ret;
```
can `ret` be a positive number smaller than `len` in case of a partial read? If it can, then we probably also should return that as an error, because that's how `struct llext_loader::read` is defined

lyakh commented on subsys/llext/shell.c:
```diff
@@ -182,10 +192,46 @@ static int cmd_llext_call_fn(const struct shell *sh, size_t argc, char *argv[])
+	res = fs_stat(argv[2], &dirent);
+	if (res) {
+		shell_error(sh, "Failed to obtain file %s, return code %d\n", argv[2], res);
+		return -ENOEXEC;
```
propagate original error code?

lyakh commented on subsys/llext/shell.c:
```diff
@@ -182,10 +192,46 @@ static int cmd_llext_call_fn(const struct shell *sh, size_t argc, char *argv[])
+		shell_print(sh, "Successfully loaded extension %s, addr %p\n", ext->name, ext);
+	} else {
+		shell_print(sh, "Failed to load extension %s, return code %d\n", ext->name, res);
+		return -ENOEXEC;
```
same with the error code? If you end up updating the PR, maybe use
```
if (res < 0) {
    shell_print(...);
    retirm res:
}
shell_print(...);
return 0;
```
similar in `llext_fs_seek()`

teburd reviewed (commented):
Looks really nice, well done! I agree with @lyakh it likely makes sense to return the error codes rather than changing them.

It would be great to see a test to ensure the feature remains functional. This subsys is under quite a lot of development from many people. Untested things tend to rot.

pillo79 commented on subsys/llext/shell.c:
```diff
@@ -11,28 +11,36 @@
 
-#define LLEXT_LOAD_HEX_HELP								\
-	"Load an elf file encoded in hex directly from the shell input. Syntax:\n"	\
+#define LLEXT_LOAD_HEX_HELP                                                                        \
+	"Load an elf file encoded in hex directly from the shell input. Syntax:\n"                 \
```
Nit: a few redundant whitespace changes in this file.

awojasinski replied:
This is how clang-format formats the code according to configuration provided in Zephyr. I decided to align all of them instead of formatting my change only

pillo79 reviewed (changes requested):
Looks very nice, only a few minor nitpicks from me on the implementation side. :+1: 

However, I disagree with the decision to add FS simulation code to the `modules` sample. That one [was designed](https://docs.zephyrproject.org/latest/samples/subsys/llext/modules/README.html) to highlight the minimal set of changes required to use LLEXT... and it is already [too difficult for newcomers to understand](https://github.com/zephyrproject-rtos/zephyr/discussions/77515). I would really prefer not to add so many things there.

I do think the code is sound, and it really looks like it could be easily adapted to run in the LLEXT test suite. Why not move the code there and add a few test cases so it runs, say, once per arch? As @teburd said, we would get continuous testing of the new API as a bonus.

awojasinski commented on include/zephyr/llext/buf_loader.h:
```diff
@@ -47,16 +47,19 @@ void *llext_buf_peek(struct llext_loader *ldr, size_t pos);
-		.buf = (_buf),				\
-		.len = (_buf_len),			\
-		.pos = 0				\
+#define LLEXT_BUF_LOADER(_buf, _buf_len)                                                           \
```
clang-format

awojasinski commented:
@teburd @pillo79 I don't fully agree with your statements. The feature is tested, samples are also executed in CI (you can see that currently two of them are failing, one in runtime and one in compilation). Moreover my change extends subsystems used in it with file system, whereas tests should examine subsystems in isolation. Nevertheless if that's your request, sure I can move it from sample and make a test case from it. Just wanted to highlight that the feature is tested

pillo79 commented:
Sure, I agree it's being compiled, however Zephyr has [a clear idea](https://docs.zephyrproject.org/latest/samples/sample_definition_and_criteria.html#samples-must-not-be-used-as-a-testcase) of what a sample should be used for... :hand_over_mouth: (same link was sent to me once as well!)  

Personally, the problem I see is more that this PR is complicating an already complex example, than that the code is not in `tests/`. But with a few additional `zasserts`, you could test all aspects of the API in addition to the basic compile test.

awojasinski committed:
llext: Add filesystem based testcase to llext test suite

Extend tests with case where extension is copied from RAM buffer
into filesystem and then loaded from filesystem.

Signed-off-by: Adam Wojasinski <awojasinski@baylibre.com>

awojasinski commented:
@pillo79 @teburd moved testing filesystem based loader from sample to tests

pillo79 reviewed (approved):
This is awesome, thank you!
